### ISSUE TITLE
--restore-database-mapping=db:db-new  and --restore-database-mapping=db-old:db return error cause substitute doesn't works 

### ISSUE THREAD

**Issue by Slach** [Altinity/clickhouse-backup#820]
```
error can't create table `db-bew`.`table`: code: 62, message: Syntax error: failed at position 20 ('-'): -new.table UUID '0b981b4a-bfac-11ee-b5ca-5271f6067e69'
```

**Comment by IvaskevychYuriy** [comment]
@Slach Hmm, I think this one should be reopened.

## TLDR; 
I've just hit this on version `2.6.5` when running `clickhouse-backup restore -t "test-db.*" --rm -s -m "test-db:test-db2" test-backup1` which fails with 
```
...
2025-02-21 21:30:57.135 INF pkg/clickhouse/clickhouse.go:1082 > CREATE TABLE ``test-db2``.My_table UUID '23138e37-f09b-11ef-aa03-361fe3815474' (`Name` String, `DateTime` DateTime) ENGINE = ReplicatedMergeTree('/clickhouse/tables/23138e37-f09b-11ef-aa03-361fe3815474/{shard}', '{replica}') PARTITION BY toDate(DateTime) ORDER BY DateTime SETTINGS index_granularity = 8192
2025-02-21 21:30:57.136 INF pkg/clickhouse/clickhouse.go:327 > clickhouse connection closed
2025-02-21 21:30:57.136 FTL cmd/clickhouse-backup/main.go:668 > error="can't create table `test-db2`.`My_table`: code: 62, message: Syntax error: failed at position 14 ('``'): ``test-db2``.My_table UUID '23138e37-f09b-11ef-aa03-361fe3815474' (`Name` String, `DateTime` DateTime) ENGINE = ReplicatedMergeTree('/clickhouse/tables/23138e37. Expected one of: IF NOT EXISTS, compound identifier, identifier after 1 times, please check your schema dependencies"
```
> NOTE: `test-db2` does not exists at first - but it gets created just fine - see Full details below. The problematic part is creating tables

Take a look at the `CREATE TABLE ``test-db2``.My_table ` part - there are double ticks (`) that cause the issue. I can confirm it works for "normal" DB names like `-m "mydb:mydb2"`. Haven't tested table mappings though - could those also have similar issue?

## Full details

### Setup & Versions
- Clickhouse cluster via Bitnami chart `7.1.4` (along with their Zookeeper). Image `bitnami/clickhouse:24.12.2-debian-12-r0`
- 2 Shards x 2 Replicas.
- Clickhouse container image with "clickhouse-client" `24.12.2.29`
- `clickhouse-backup` from image `altinity/clickhouse-backup:2.6.5` running in Clickhouse pods as a sidecar
>  Version:         2.6.5
    Git Commit:      68dcef9c2cfe1d88050082d6f56654156f819538
    Build Date:      2024-12-31

### Steps to reproduce
1. Exec into clickhouse container (1st shard 1st replica)
2. Create database:
```
I have no name!@clickhouse-shard0-0:/$ clickhouse-client --user "$CLICKHOUSE_ADMIN_USER" --password "$CLICKHOUSE_ADMIN_PASSWORD" -q "CREATE DATABASE IF NOT EXISTS \"test-db\" ON CLUSTER cluster ENGINE = Replicated('/clickhouse/databases/tes
t-db', '{replica}');"
clickhouse-shard0-0.clickhouse-headless.infra.svc.cluster.local 9000    0               3       0
clickhouse-shard1-1.clickhouse-headless.infra.svc.cluster.local 9000    0               2       0
clickhouse-shard1-0.clickhouse-headless.infra.svc.cluster.local 9000    0               1       0
clickhouse-shard0-1.clickhouse-headless.infra.svc.cluster.local 9000    0               0       0
```
3. Create table:
```
I have no name!@clickhouse-shard0-0:/$ clickhouse-client --user "$CLICKHOUSE_ADMIN_USER" --password "$CLICKHOUSE_ADMIN_PASSWORD" -q "CREATE TABLE IF NOT EXISTS \"test-db\".My_table (Name String, DateTime DateTime) ENGINE = ReplicatedMergeTree() PARTITION BY (toDate(DateTime)) ORDER BY (DateTime)"
clickhouse-shard0-0     clickhouse-shard0-0     OK      3       3
clickhouse-shard0-1     clickhouse-shard0-1     OK      2       1
clickhouse-shard1-0     clickhouse-shard1-0     OK      1       1
clickhouse-shard1-1     clickhouse-shard1-1     OK      0       0
```
4. Now exec into sidecar container with backup CLI
5. Create backup:
```
clickhouse-shard0-0:/$ clickhouse-backup create -t "test-db.*" test-backup1
2025-02-21 21:28:17.633 INF pkg/clickhouse/clickhouse.go:130 > clickhouse connection prepared: tcp://localhost:9000 run ping
2025-02-21 21:28:17.637 INF pkg/clickhouse/clickhouse.go:133 > clickhouse connection success: tcp://localhost:9000
2025-02-21 21:28:17.637 INF pkg/clickhouse/clickhouse.go:1082 > SELECT metadata_path FROM system.tables WHERE database = 'system' AND metadata_path!='' LIMIT 1;
2025-02-21 21:28:17.641 INF pkg/clickhouse/clickhouse.go:1082 > SELECT name, engine FROM system.databases WHERE NOT match(name,'^(system|INFORMATION_SCHEMA|information_schema|_temporary_and_external_tables|signoz_.*)$') AND match(name,'^(test-db)$')
2025-02-21 21:28:17.643 INF pkg/clickhouse/clickhouse.go:1082 > SHOW CREATE DATABASE `test-db`
2025-02-21 21:28:17.645 INF pkg/clickhouse/clickhouse.go:1080 > SELECT name, count(*) as is_present FROM system.settings WHERE name IN (?, ?) GROUP BY name with args [show_table_uuid_in_table_create_query_if_not_nil display_secrets_in_show_and_select]
2025-02-21 21:28:17.649 INF pkg/clickhouse/clickhouse.go:1082 > SELECT name FROM system.databases WHERE engine IN ('MySQL','PostgreSQL','MaterializedPostgreSQL')
2025-02-21 21:28:17.651 INF pkg/clickhouse/clickhouse.go:1082 >    SELECT     countIf(name='data_path') is_data_path_present,     countIf(name='data_paths') is_data_paths_present,     countIf(name='uuid') is_uuid_present,     countIf(name='create_table_query') is_create_table_query_present,     countIf(name='total_bytes') is_total_bytes_present    FROM system.columns WHERE database='system' AND table='tables'
2025-02-21 21:28:17.655 INF pkg/clickhouse/clickhouse.go:1082 > SELECT database, name, engine , data_paths , uuid , create_table_query , coalesce(total_bytes, 0) AS total_bytes   FROM system.tables WHERE is_temporary = 0 AND match(concat(database,'.',name),'^test\-db\..*$')  ORDER BY total_bytes DESC SETTINGS show_table_uuid_in_table_create_query_if_not_nil=1
2025-02-21 21:28:17.732 INF pkg/clickhouse/clickhouse.go:1082 > SELECT metadata_path FROM system.tables WHERE database = 'system' AND metadata_path!='' LIMIT 1;
2025-02-21 21:28:17.735 INF pkg/clickhouse/clickhouse.go:1082 > SELECT sum(bytes_on_disk) as size FROM system.parts WHERE active AND database='test-db' AND table='My_table' GROUP BY database, table
2025-02-21 21:28:17.738 INF pkg/clickhouse/clickhouse.go:1082 > SELECT count() as cnt FROM system.columns WHERE database='system' AND table='functions' AND name='create_query' SETTINGS empty_result_for_aggregation_by_empty_set=0
2025-02-21 21:28:17.741 INF pkg/clickhouse/clickhouse.go:1082 > SELECT name, create_query FROM system.functions WHERE create_query!=''
2025-02-21 21:28:17.746 INF pkg/clickhouse/clickhouse.go:1082 > SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER'
2025-02-21 21:28:17.747 INF pkg/clickhouse/clickhouse.go:1082 > SELECT countIf(name='type') AS is_disk_type_present, countIf(name='object_storage_type') AS is_object_storage_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')
2025-02-21 21:28:17.751 INF pkg/clickhouse/clickhouse.go:1082 > SELECT d.path AS path, any(d.name) AS name, any(lower(if(d.type='ObjectStorage',d.object_storage_type,d.type))) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  LEFT JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path
2025-02-21 21:28:17.762 INF pkg/clickhouse/clickhouse.go:1082 > SELECT name FROM system.user_directories WHERE type='replicated'
2025-02-21 21:28:17.763 INF pkg/clickhouse/clickhouse.go:1082 > SELECT JSONExtractString(params,'path') AS access_path FROM system.user_directories WHERE type='local directory'
2025-02-21 21:28:17.765 INF pkg/clickhouse/clickhouse.go:1082 > SELECT metadata_path FROM system.tables WHERE database = 'system' AND metadata_path!='' LIMIT 1;
2025-02-21 21:28:17.773 INF pkg/backup/create.go:177 > done createBackupRBAC size=2.84KiB
2025-02-21 21:28:17.773 INF pkg/clickhouse/clickhouse.go:1080 > SELECT column, groupUniqArray(type) AS uniq_types FROM system.parts_columns WHERE active AND database=? AND table=? AND type NOT LIKE 'Enum%!((MISSING)%!'(MISSING) AND type NOT LIKE 'Tuple(%!'(MISSING) AND type NOT LIKE 'Nullable(Enum%!((MISSING)%!'(MISSING) AND type NOT LIKE 'Nullable(Tuple(%!'(MISSING) AND type NOT LIKE 'Array(Tuple(%!'(MISSING) AND type NOT LIKE 'Nullable(Array(Tuple(%!'(MISSING) GROUP BY column HAVING length(uniq_types) > 1 with args [test-db My_table]
2025-02-21 21:28:17.777 INF pkg/clickhouse/clickhouse.go:1082 > ALTER TABLE `test-db`.`My_table` FREEZE WITH NAME 'a16a98c09c394d72961da9af07d90c26';
2025-02-21 21:28:17.800 INF pkg/clickhouse/clickhouse.go:1082 > ALTER TABLE `test-db`.`My_table` UNFREEZE WITH NAME 'a16a98c09c394d72961da9af07d90c26'
2025-02-21 21:28:17.801 INF pkg/clickhouse/clickhouse.go:1080 > SELECT mutation_id, command FROM system.mutations WHERE is_done=0 AND database=? AND table=? with args [test-db My_table]
2025-02-21 21:28:17.806 INF pkg/backup/create.go:344 > done progress=1/1 table=test-db.My_table
2025-02-21 21:28:17.806 INF pkg/clickhouse/clickhouse.go:1082 > SELECT value FROM `system`.`build_options` WHERE name='VERSION_DESCRIBE'
2025-02-21 21:28:17.809 INF pkg/backup/create.go:356 > done duration=176ms operation=createBackupLocal version=2.6.5
2025-02-21 21:28:17.810 INF pkg/clickhouse/clickhouse.go:130 > clickhouse connection prepared: tcp://localhost:9000 run ping
2025-02-21 21:28:17.811 INF pkg/clickhouse/clickhouse.go:133 > clickhouse connection success: tcp://localhost:9000
2025-02-21 21:28:17.811 INF pkg/clickhouse/clickhouse.go:1082 > SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0
2025-02-21 21:28:17.815 INF pkg/clickhouse/clickhouse.go:1082 > SELECT macro, substitution FROM system.macros
2025-02-21 21:28:17.817 INF pkg/clickhouse/clickhouse.go:1082 > SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0
2025-02-21 21:28:17.821 INF pkg/clickhouse/clickhouse.go:1082 > SELECT macro, substitution FROM system.macros
2025-02-21 21:28:17.930 INF pkg/storage/general.go:163 > list_duration=33.720324
2025-02-21 21:28:17.930 INF pkg/backup/delete.go:157 > remove '/bitnami/clickhouse/data/backup/2025-02-21T21-09-39'
2025-02-21 21:28:17.930 INF pkg/backup/delete.go:166 > done backup=2025-02-21T21-09-39 duration=120ms location=local operation=delete
2025-02-21 21:28:17.930 INF pkg/clickhouse/clickhouse.go:327 > clickhouse connection closed
2025-02-21 21:28:17.930 INF pkg/clickhouse/clickhouse.go:327 > clickhouse connection closed
```
6. Ensure it exists locally:
```
clickhouse-shard0-0:/$ clickhouse-backup list local
2025-02-21 21:29:13.149 INF pkg/clickhouse/clickhouse.go:130 > clickhouse connection prepared: tcp://localhost:9000 run ping
2025-02-21 21:29:13.151 INF pkg/clickhouse/clickhouse.go:133 > clickhouse connection success: tcp://localhost:9000
2025-02-21 21:29:13.151 INF pkg/clickhouse/clickhouse.go:1082 > SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER'
2025-02-21 21:29:13.153 INF pkg/clickhouse/clickhouse.go:1082 > SELECT countIf(name='type') AS is_disk_type_present, countIf(name='object_storage_type') AS is_object_storage_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')
2025-02-21 21:29:13.158 INF pkg/clickhouse/clickhouse.go:1082 > SELECT d.path AS path, any(d.name) AS name, any(lower(if(d.type='ObjectStorage',d.object_storage_type,d.type))) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  LEFT JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path
test-backup1   3.22KiB   21/02/2025 21:28:17   local      regular
2025-02-21 21:29:13.165 INF pkg/clickhouse/clickhouse.go:327 > clickhouse connection closed
``` 
7. Attempt to restore:
```
clickhouse-shard0-0:/$ clickhouse-backup restore -t "test-db.*" --rm -s -m "test-db:test-db2" test-backup1
2025-02-21 21:30:57.058 INF pkg/clickhouse/clickhouse.go:130 > clickhouse connection prepared: tcp://localhost:9000 run ping
2025-02-21 21:30:57.062 INF pkg/clickhouse/clickhouse.go:133 > clickhouse connection success: tcp://localhost:9000
2025-02-21 21:30:57.062 INF pkg/clickhouse/clickhouse.go:1082 > SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER'
2025-02-21 21:30:57.065 INF pkg/clickhouse/clickhouse.go:1082 > SELECT countIf(name='type') AS is_disk_type_present, countIf(name='object_storage_type') AS is_object_storage_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')
2025-02-21 21:30:57.069 INF pkg/clickhouse/clickhouse.go:1082 > SELECT d.path AS path, any(d.name) AS name, any(lower(if(d.type='ObjectStorage',d.object_storage_type,d.type))) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  LEFT JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path
2025-02-21 21:30:57.085 INF pkg/clickhouse/clickhouse.go:1082 > SELECT engine FROM system.databases WHERE name = 'test-db2'
2025-02-21 21:30:57.088 INF pkg/clickhouse/clickhouse.go:1082 > DROP TABLE IF EXISTS `test-db2`.`My_table`
2025-02-21 21:30:57.093 INF pkg/clickhouse/clickhouse.go:1082 > CREATE DATABASE IF NOT EXISTS `test-db2`
2025-02-21 21:30:57.120 INF pkg/clickhouse/clickhouse.go:1082 > SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0
2025-02-21 21:30:57.124 INF pkg/clickhouse/clickhouse.go:1082 > SELECT macro, substitution FROM system.macros
2025-02-21 21:30:57.126 INF pkg/clickhouse/clickhouse.go:1082 > SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0
2025-02-21 21:30:57.128 INF pkg/clickhouse/clickhouse.go:1082 > SELECT macro, substitution FROM system.macros
2025-02-21 21:30:57.130 INF pkg/clickhouse/clickhouse.go:1080 > SELECT count() FROM system.zookeeper WHERE path=? with args [/clickhouse/tables/23138e37-f09b-11ef-aa03-361fe3815474/shard0/replicas/clickhouse-shard0-0]
2025-02-21 21:30:57.135 INF pkg/clickhouse/clickhouse.go:1082 > CREATE TABLE ``test-db2``.My_table UUID '23138e37-f09b-11ef-aa03-361fe3815474' (`Name` String, `DateTime` DateTime) ENGINE = ReplicatedMergeTree('/clickhouse/tables/23138e37-f09b-11ef-aa03-361fe3815474/{shard}', '{replica}') PARTITION BY toDate(DateTime) ORDER BY DateTime SETTINGS index_granularity = 8192
2025-02-21 21:30:57.136 INF pkg/clickhouse/clickhouse.go:327 > clickhouse connection closed
2025-02-21 21:30:57.136 FTL cmd/clickhouse-backup/main.go:668 > error="can't create table `test-db2`.`My_table`: code: 62, message: Syntax error: failed at position 14 ('``'): ``test-db2``.My_table UUID '23138e37-f09b-11ef-aa03-361fe3815474' (`Name` String, `DateTime` DateTime) ENGINE = ReplicatedMergeTree('/clickhouse/tables/23138e37. Expected one of: IF NOT EXISTS, compound identifier, identifier after 1 times, please check your schema dependencies"
```

### Expected result
Tables are created successfully in new empty Database which also gets created if not exist yet.

### Actual result
Observe the error on creating tables. Database is created just fine though


Let me know if I can provide any other info\debug

**Comment by Slach** [comment]
thanks for detailed report

**Comment by IvaskevychYuriy** [comment]
@Slach Hello again :) So I've just upgraded from `2.6.5` to `2.6.11` to test out the fix and it does work for database with special symbols in the name - thanks!

However, it fails when there's a _table_ with special symbols in the name, like `t1-d`. Notice from below log that ``CREATE TABLE `test-db2`.t1-d`` is executed without escaping table name, it should be ``CREATE TABLE `test-db2`.`t1-d` ``:

```
...
2025-04-02 18:25:42.095 INF pkg/clickhouse/clickhouse.go:1123 > CREATE TABLE `test-db2`.t1-d UUID 'e2737ad0-0fef-11f0-acaa-9277ff4e297c' (`DateTime` DateTime, `Id` UUID, `Name` String) ENGINE = Distributed('cluster', 'test-db2', 't1', sipHash64(toDate(DateTime)))
2025-04-02 18:25:42.096 WRN pkg/backup/restore.go:1169 > can't create table 'test-db2.t1-d': code: 62, message: Syntax error: failed at position 27 ('-'): -d UUID 'e2737ad0-0fef-11f0-acaa-9277ff4e297c' (`DateTime` DateTime, `Id` UUID, `Name` String) ENGINE = Distributed('cluster', 'test-db2', 't1', sipHash64(toDat. Expected one of: ParserArrayOfJSONIdentifierDelimiter, token sequence, OpeningSquareBracket, Dot, token, UUID, ON, OpeningRoundBracket, storage definition, ENGINE, PARTITION BY, PRIMARY KEY, ORDER BY, SAMPLE BY, TTL, EMPTY AS, CLONE AS, AS, COMMENT, INTO OUTFILE, FORMAT, SETTINGS, end of query, will try again
```

Steps to reproduce from above are still relevant, this reproduces on `clickhouse-backup restore -t "test-db.*" --rm -s -m "test-db:test-db2" test-backup1`
The only change is to create table in step 3 like this:
```sql
-- this restores ok
CREATE TABLE IF NOT EXISTS `t1`
(
    DateTime DateTime,
    Id UUID,
    Name String
) 
ENGINE = ReplicatedMergeTree()
PARTITION BY toDate(DateTime)
ORDER BY DateTime

-- this fails
CREATE TABLE IF NOT EXISTS `t1-d` AS `t1`
ENGINE = Distributed('cluster', 'test-db', 't1', sipHash64(toDate(DateTime)));
```

Whether to reopen this issue one or create a new one is up to you

**Comment by Slach** [comment]
fixed in 2.6.13
